pkg/controllers/user/pipeline/controller/pipelineexecution/registrycertsyncer.go

```diff
@@ -0,0 +1,186 @@
+package pipelineexecution
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/cert"
+	"time"
+)
+
+// This Syncer is responsible for certificates rotation for internal registries
+
+const (
+	checkCertRotateInterval = 12 * time.Hour
+	//rotateThreshold set the rotation threshold so it will rotate at 90% of the total lifetime of the certificates.
+	rotateThreshold = 0.9
+	maxRetry        = 60
+)
+
+type RegistryCertSyncer struct {
+	clusterName string
+
+	pods                    v1.PodInterface
+	podLister               v1.PodLister
+	secrets                 v1.SecretInterface
+	secretLister            v1.SecretLister
+	managementSecretLister  v1.SecretLister
+	namespaceLister         v1.NamespaceLister
+	pipelineExecutionLister v3.PipelineExecutionLister
+	pipelineSettingLister   v3.PipelineSettingLister
+}
+
+func (s *RegistryCertSyncer) sync(ctx context.Context, syncInterval time.Duration) {
+	for range ticker.Context(ctx, checkCertRotateInterval) {
+		s.checkAndRotateCerts(ctx)
+	}
+}
+
+func (s *RegistryCertSyncer) checkAndRotateCerts(ctx context.Context) {
+	labelsSearchSet := labels.Set{utils.PipelineNamespaceLabel: "true"}
+	namespaces, err := s.namespaceLister.List("", labels.SelectorFromSet(labelsSearchSet))
+	if err != nil {
+		logrus.Error(err)
+		return
+	}
+	for _, ns := range namespaces {
+		secret, err := s.secretLister.Get(ns.Name, utils.RegistryCrtSecretName)
+		if apierrors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logrus.Error(err)
+			return
+		}
+		if !s.shouldRotate(secret.Data[utils.RegistryCrt]) {
+			continue
+		}
+		projectID := getProjectID(ns)
+		if projectID != "" {
+			go func() {
+				if err := s.rotateCerts(ctx, projectID); err != nil {
+					errors.Wrapf(err, "fail to rotate registry certs for %s project", projectID)
```

Comment:
Just want to make sure we are ok with the fact that if the cert rotation fails, the next retry will be only on the next sync

Reply:
Another chance is that when the certs configmap is missing on an execution, it will be re-generated. So if the rotation fails and users find that certs become invalid before next retry sync, they can delete the configmap to get a new one as the workaround.